build-info-extractor/src/main/java/org/jfrog/build/extractor/executor/CommandExecutor.java

```diff
@@ -123,28 +124,53 @@ static String maskCredentials(String command, List<String> credentials) {
+    }
+
+    /**
+     * Execute a command in external process.
+     *
+     * @param execDir     - The execution dir (Usually path to project). Null means current directory.
+     * @param args        - Command arguments.
+     * @param credentials - If specified, the credentials will be concatenated to the other commands. The credentials will be masked in the log output.
+     * @param logger      - The logger which will log the running command.
+     * @param timeout     - The maximum time to wait for the command execution.
+     * @param unit        – The time unit of the timeout argument.
+     * @return CommandResults object
+     */
+    public CommandResults exeCommand(File execDir, List<String> args, List<String> credentials, Log logger, long timeout, TimeUnit unit) throws InterruptedException, IOException {
         args.add(0, executablePath);
         ExecutorService service = Executors.newFixedThreadPool(2);
         try {
             Process process = runProcess(execDir, args, credentials, env, logger);
             // The output stream is not necessary in non-interactive scenarios, therefore we can close it now.
             process.getOutputStream().close();
-            try (InputStream inputStream = process.getInputStream();
-                 InputStream errorStream = process.getErrorStream()) {
+            try (InputStream inputStream = process.getInputStream(); InputStream errorStream = process.getErrorStream()) {
                 StreamReader inputStreamReader = new StreamReader(inputStream);
                 StreamReader errorStreamReader = new StreamReader(errorStream);
                 service.submit(inputStreamReader);
                 service.submit(errorStreamReader);
-                process.waitFor();
+                boolean executionTerminatedProperly = process.waitFor(timeout, unit);
                 service.shutdown();
-                boolean terminatedProperly = service.awaitTermination(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+                boolean outputReaderTerminatedProperly = service.awaitTermination(READER_SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+                boolean terminatedProperly = executionTerminatedProperly && outputReaderTerminatedProperly;
                 return getCommandResults(terminatedProperly, args, inputStreamReader.getOutput(), errorStreamReader.getOutput(), process.exitValue());
+            } finally {
+                // Ensure termination of the subprocess we have created.
+                if (process.isAlive()) {
+                    // First try to terminate the process gracefully
+                    process.destroy();
+                    process.waitFor(PROCESS_TERMINATION_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+                    if (process.isAlive()) {
+                        // Finally, force kill.
+                        process.destroyForcibly();
+                    }
+                }
```

Comment:
Instead of closing the readers before the process termination, it may be better to close them after the process finishes. This way, we ensure that all output generated by the process is captured, including any output that might occur following its termination.
What do you think?